File: chat/src/main/java/org/thoughtcrime/securesms/dependencies/ApplicationDependencyProvider.java
package org.thoughtcrime.securesms.dependencies;

import android.app.Application;

import androidx.annotation.NonNull;

import org.thoughtcrime.securesms.database.JobDatabase;
import org.thoughtcrime.securesms.jobmanager.JobManager;
import org.thoughtcrime.securesms.jobmanager.impl.FactoryJobPredicate;
import org.thoughtcrime.securesms.jobmanager.impl.JsonDataSerializer;
import org.thoughtcrime.securesms.jobs.DownloadAttachmentJob;
import org.thoughtcrime.securesms.jobs.FastJobStorage;
import org.thoughtcrime.securesms.jobs.JobManagerFactories;
import org.thoughtcrime.securesms.video.exo.SimpleExoPlayerPool;
import com.difft.android.websocket.api.SignalServiceAccountManager;
import com.difft.android.websocket.internal.configuration.SignalServiceConfiguration;

/**
 * Implementation of {@link ApplicationDependencies.Provider} that provides real app dependencies.
 */
public class ApplicationDependencyProvider implements ApplicationDependencies.Provider {

    private final Application context;

    public ApplicationDependencyProvider(@NonNull Application context) {
        this.context = context;
    }


    @NonNull
    @Override
    public JobManager provideJobManager() {
        JobManager.Configuration config = new JobManager.Configuration.Builder()
                .setDataSerializer(new JsonDataSerializer())
                .setJobFactories(JobManagerFactories.getJobFactories(context))
                .setConstraintFactories(JobManagerFactories.getConstraintFactories(context))
                .setConstraintObservers(JobManagerFactories.getConstraintObservers(context))
                .setJobStorage(new FastJobStorage(JobDatabase.getInstance(context)))
                //.setJobMigrator(new JobMigrator(TextSecurePreferences.getJobManagerVersion(context), JobManager.CURRENT_VERSION, JobManagerFactories.getJobMigrations(context)))
                .addReservedJobRunner(new FactoryJobPredicate(DownloadAttachmentJob.KEY))
                .build();
        return new JobManager(context, config);
    }

    @Override
    public @NonNull
    SignalServiceAccountManager provideSignalServiceAccountManager(@NonNull SignalServiceConfiguration signalServiceConfiguration) {
        return new SignalServiceAccountManager(signalServiceConfiguration, true);
    }


    @NonNull
    @Override
    public SimpleExoPlayerPool provideExoPlayerPool() {
        return new SimpleExoPlayerPool(context);
    }
}

File: chat/src/main/java/com/difft/android/chat/dependencies/ApplicationDependencyProvider.kt
package com.difft.android.chat.dependencies

import android.app.Application
import com.difft.android.chat.jobmanager.JobManager
import com.difft.android.chat.jobmanager.impl.FactoryJobPredicate
import com.difft.android.chat.jobmanager.impl.JsonDataSerializer
import com.difft.android.chat.jobs.DownloadAttachmentJob
import com.difft.android.chat.jobs.FastJobStorage
import com.difft.android.chat.jobs.JobManagerFactories
import com.difft.android.chat.jobs.WcdbJobStorage
import com.difft.android.chat.video.exo.SimpleExoPlayerPool
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent

/**
 * Implementation of [ApplicationDependencies.Provider] that provides real app dependencies.
 */
class ApplicationDependencyProvider(
    private val context: Application
) : ApplicationDependencies.Provider {

    /**
     * Hilt EntryPoint used to obtain the [WcdbJobStorage] singleton from a
     * non-Hilt construction site. [ApplicationDependencyProvider] is built by
     * hand from `TempTalkApplication.onCreate` before Hilt injection reaches
     * this class, so we bridge via `EntryPointAccessors` — same pattern already
     * used for `MessageStore` / `EnvironmentHelper` elsewhere in the codebase
     * (design §4.2.3, D16).
     */
    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface DepsEntryPoint {
        fun wcdbJobStorage(): WcdbJobStorage
    }

    override fun provideJobManager(): JobManager {
        val wcdbJobStorage = EntryPointAccessors
            .fromApplication(context, DepsEntryPoint::class.java)
            .wcdbJobStorage()
        val config = JobManager.Configuration.Builder()
            // NOTE: ChatHiltProvidesModule.provideJobDataSerializer() provides a Hilt-bound twin
            // of this serializer for ReactionSendCoordinator. Keep in sync if this line changes.
            .setDataSerializer(JsonDataSerializer())
            .setJobFactories(JobManagerFactories.getJobFactories(context))
            .setConstraintFactories(JobManagerFactories.getConstraintFactories(context))
            .setConstraintObservers(JobManagerFactories.getConstraintObservers(context))
            .setJobStorage(FastJobStorage(wcdbJobStorage))
            .addReservedJobRunner(FactoryJobPredicate(DownloadAttachmentJob.KEY))
            .build()
        return JobManager(context, config)
    }

    override fun provideExoPlayerPool(): SimpleExoPlayerPool {
        return SimpleExoPlayerPool(context)
    }
}
